### PR TITLE
Add backend for job cancellation.

### PR DESCRIPTION
**Pull request by ncik-roberts**
Just adds a cancel endpoint that the client can call. Then, we try to kill all autodriver processes on the worker.

I mucked around with trying to cull some duplication in directory names in `tangoREST.py`, but it didn't end up helping too much.

### CODE REVIEW

File: tango.py [hunk 1]
```diff
@@ -109,9 +109,74 @@ def delJob(self, id, deadjob):
...
+
+        if job_status == JobQueue.JobStatus.NOT_FOUND:
+            return -1
+        elif job_status == JobQueue.JobStatus.DEAD:
+            return -2
+        elif job_status == JobQueue.JobStatus.RUNNING:
+            return self.killUntilJobComplete(id, job)
+        else:
+           assert job_status == JobQueue.JobStatus.WAITING
+           # In this case, findRemovingLive has moved the live job to the dead
```
Comment (jburns20) [inline tango.py:156]: findRemovingWaiting, right?

File: vmms/ec2SSH.py [hunk 2]
```diff
@@ -411,11 +411,19 @@ def copyIn(self, vm, inputFiles):
...
         return 0
 
+    def sshWithTimeout(self, vm, runcmd, runTimeout):
+        domain_name = self.domainName(vm)
+        account = "%s@%s" % (config.Config.EC2_USER_NAME, domain_name)
+        return timeout(["ssh"] + self.ssh_flags + [account] + runcmd, runTimeout)
+
+    def kill(self, vm, runTimeout):
+        self.log.debug("pkill: Killing job on VM %s" % self.instanceName(vm.id, vm.name))
+        return self.sshWithTimeout(vm, ["/usr/bin/killall", "autodriver"], runTimeout)
```
Comment (jburns20) [inline vmms/ec2SSH.py]: So I know I said that we could do it this way, but after further investigation, it appears that we also need to run the cleanup() function of the autodriver in order to really make sure everything is gone and the instance is ready for another job. (See https://github.com/CMU-15-411-F18/Tango/blob/master/autodriver/autodriver.c#L625).

My initial thought on this is that we could add a new command-line argument to the autodriver that would cause it to just run cleanup() instead of actually running the job, but I'm open to other ideas.

File: restful-tango/tangoREST.py [hunk 3]
```diff
@@ -130,15 +139,14 @@ def createTangoMachine(self, image, vmms=Config.VMMS_NAME,
             disk=None,
             network=None)
 
-    def convertJobObj(self, dirName, jobObj):
+    def convertJobObj(self, key, courselab, jobObj):
```
Comment (jburns20) [inline restful-tango/tangoREST.py:143]: What's the reason for this change? It seems unrelated to job canceling.

File: vmms/ec2SSH.py (hunk 2 above)
Comment (jburns20) [reply vmms/ec2SSH.py]: The other issue is that in my testing, the current method doesn't seem to actually kill the autodriver at all. I added a `sleep 60` to the Makefile, clicked cancel, and the page didn't load until the job finished. After the job finished normally, the page loaded and I got an error saying that canceling the job failed.

**Review by jburns20 — CHANGES_REQUESTED** [review]

File: vmms/ec2SSH.py (hunk 2 above)
Comment (jburns20) [reply vmms/ec2SSH.py]: Here's another idea: have the autodriver respond to a SIGINT and immediately call cleanup() from there.

Comment (ncik-roberts) [reply vmms/ec2SSH.py]: I think I'll go for the latter option (responding to SIGINT).

File: restful-tango/tangoREST.py (hunk 3 above)
Comment (ncik-roberts) [reply restful-tango/tangoREST.py:143]: I got annoyed by the 1000 different ways of constructing the path so I refactored the code in this file a bit. I think I'll roll it back just so we're more assured that it works, haha.

File: tango.py [hunk 4]
```diff
@@ -109,9 +109,74 @@ def delJob(self, id, deadjob):
...
+          1. The job with the specified output file does not exist
+          2. the job with the specified output file has finished running normally
+          3. The job with the specified output file has been cancelled
+          4. The job was found, and it's running, but cancellation failed.
+        In case 1, -1 is returned.
+                2, -2 is returned.
+                3,  0 is returned.
+                4, -3 is returned.
+        """
+        self.log.error("Received cancelJobWithPath(%s) request" % (outFilePath))
```
Comment (jburns20) [inline tango.py]: This should be self.log.debug.

File: autodriver/autodriver.c [hunk 5]
```diff
@@ -618,6 +619,15 @@ static void cleanup(void) {
     }
 }
 
+/**
+ * @brief Signal handler--compatible signal declaration.
+ */
+static void cleanup_hndlr(int sig) {
+    (void) sig; // This puts sig into the void.
```
Comment (jburns20) [inline autodriver/autodriver.c:663]: you might just be able to remove the argument entirely?

Comment (ncik-roberts) [reply autodriver/autodriver.c:663]: I was getting warnings if I passed just &cleanup to signal. I am going to put logging into cleanup_hndlr anyway.

**Review by jburns20 — APPROVED** [review]
